### ISSUE TITLE
ngTemplateOutletContext changes ignored in outputs

### ISSUE THREAD

**Issue by Eddman** [angular/angular#42832]
### Which @angular/* package(s) are the source of the bug?

common

### Is this a regression?

Yes

### Description

We have an input inside of our table that uses `[(ngModel)]`. The input is rendered using `ngTemplateOutlet`. When we reload data in the table the `(ngModelChange)` binding stop working. The problem is in the way how Angular works with context:
In `NgTemplateOutlet`:
```typescript
ngOnChanges(changes: SimpleChanges) {
  if (changes['ngTemplateOutlet']) {
    const viewContainerRef = this._viewContainerRef;

    if (this._viewRef) {
      viewContainerRef.remove(viewContainerRef.indexOf(this._viewRef));
    }

    this._viewRef = this.ngTemplateOutlet ?
        viewContainerRef.createEmbeddedView(this.ngTemplateOutlet, this.ngTemplateOutletContext) :
        null;
  } else if (
      this._viewRef && changes['ngTemplateOutletContext'] && this.ngTemplateOutletContext) {
    this._viewRef.context = this.ngTemplateOutletContext;  <----- the context is correctly set on change
  }
}
```
But in our generated code:
```typescript
function DerivativesComponent_risk_data_table_column_121_ng_template_1_Template(rf, 
  ctx) { if (rf & 1) { <---- the initial context is used as parameter
    _angular_core__WEBPACK_IMPORTED_MODULE_19__["ɵɵelementStart"](0, "mat-form-field", 3);
    _angular_core__WEBPACK_IMPORTED_MODULE_19__["ɵɵelementStart"](1, "input", 67, 68);
    _angular_core__WEBPACK_IMPORTED_MODULE_19__["ɵɵlistener"]("ngModelChange", 
      function DerivativesComponent_risk_data_table_column_121_ng_template_1_Template_input_ngModelChange_1_listener($event) {
         const row_r67 = ctx.row; return row_r67.localPrices.settlementPrice = $event; });
```
So the listener is using always the original context unless a new `createEmbeddedView` is called. The generated template function is not called on context set.

### Please provide the environment you discovered this bug in

```true
Angular CLI: 12.1.1
Node: 14.16.0
Package Manager: npm 6.14.13
OS: linux x64

Angular: 12.1.1
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, language-service, material, platform-browser
... platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1201.1
@angular-devkit/build-angular   12.1.1
@angular-devkit/core            12.1.1
@angular-devkit/schematics      12.1.1
@schematics/angular             12.1.1
rxjs                            6.6.7
typescript                      4.2.4
```

### Anything else?

I've marked this bug as regression, because I've reverted our code to Angular 11 and there it worked. The bug can be observed in all Angular 12 versions. 

Looking at the code this was introduced in #40360.

**Comment by Eddman** [comment]
Just wanted to add that if we switch to
```
(ngModelChange)="priceInputModelChanged(row, $event)"
```
The generated code changes to:
```
_angular_core__WEBPACK_IMPORTED_MODULE_19__["ɵɵlistener"]("ngModelChange", function DerivativesComponent_risk_data_table_column_121_ng_template_1_Template_input_ngModelChange_1_listener($event) { const restoredCtx = _angular_core__WEBPACK_IMPORTED_MODULE_19__["ɵɵrestoreView"](_r72); const row_r67 = restoredCtx.row; const ctx_r71 = _angular_core__WEBPACK_IMPORTED_MODULE_19__["ɵɵnextContext"](2); return ctx_r71.priceInputModelChanged(row_r67, $event); });
```
So it uses `const restoredCtx = _angular_core__WEBPACK_IMPORTED_MODULE_19__["ɵɵrestoreView"](_r72);` and with that it works just fine.

The expressions
```
[(ngModel)]="row.localPrices.settlementPrice"
```
or
```
(ngModelChange)="row.localPrices.settlementPrice = $event"
```
generate the function without `restoredCtx`.

**Comment by JoostK** [comment]
Thanks for the _excellent_ issue report! This was also reported in #42698 and fixed in #42755 which will be in this week's release.
